velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -135,79 +125,62 @@ struct RegexpReplaceFunction {
       const arg_type<Varchar>& pattern,
       const arg_type<Varchar>& replace,
       const arg_type<int64_t>& position) {
-    VELOX_USER_CHECK_GE(position, 1, "regexp_replace requires a position >= 1");
-
-    re2::RE2* patternRegex = getCachedRegex(pattern.str());
-    re2::StringPiece replaceStringPiece = toStringPiece(replace);
-    re2::StringPiece inputStringPiece = toStringPiece(stringInput);
-
-    if (position > stringInput.size() + 1) {
-      result.resize(inputStringPiece.size());
-      std::memcpy(
-          result.data(), inputStringPiece.data(), inputStringPiece.size());
+    if (position > int(stringInput.size()) + 1) {
+      result = stringInput;
       return;
     }
 
-    // Adjust the position for UTF-8 by counting the code points.
-    size_t utf8Position = 0;
-    size_t numCodePoints = 0;
-    while (numCodePoints < position - 1 && utf8Position <= stringInput.size()) {
-      int charLength =
-          utf8proc_char_length(inputStringPiece.data() + utf8Position);
-      VELOX_USER_CHECK_GT(
-          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
-      ++numCodePoints;
-      utf8Position += charLength;
+    VELOX_USER_CHECK_GE(position, 1, "regexp_replace requires a position >= 1");
+    if (stringInput.size() == 0) {
+      if (pattern.size() == 0 && position == 1) {
+        result = replace;
+        return;
+      }
+      if (pattern.size() > 0) {
+        result = stringInput;
+        return;
+      }
     }
+
+    size_t utf8Position =
```

Comment:
nit: utf8Position -> start

If you look at usages of cappedLength, you'll notice that it has a template parameter isAscii. This is because ASCII strings are a lot cheaper to process than UTF-8 and there are a lot of ASCII-only datasets.

I suggest to model this function the same way as cappedLength and add callAscii here. See https://facebookincubator.github.io/velox/develop/scalar-functions.html#all-ascii-fast-path